From 10a630656fadb160ec3e01b934adac3e1ca600f9 Mon Sep 17 00:00:00 2001
From: Ray Johnston <ray.johnston@artifex.com>
Date: Sun, 23 Dec 2018 13:02:22 -0800
Subject: [PATCH 1/1] Bug 700359: Uniprint device did not set
 color_info.polarity needed by pdf14

The pdf14 device assumes that if the polarity is ADDITIVE, the colorspace
must be Gray or RGB. The uniprint device did not set the polarity when the
colorspace was changed to CMYK, so the number of components set in the pdf14
compositor was 3, even though the ICC profile for the device expects 4 as
num_components in. This caused a mismatch when reading the clist because
it expected 4 components and only 3 were written. Note that the RGBW, RGBOV,
and RGBNOV modes will probably still fail, but unti we see a case for a device
needing these unusual modes, just leave it as it has been for many years.

Note that other devices with strange colorspaces, such as RGBA may also
trip over this.
---
 devices/gdevupd.c | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/devices/gdevupd.c b/devices/gdevupd.c
index 6d50f14..22353a4 100644
--- a/devices/gdevupd.c
+++ b/devices/gdevupd.c
@@ -1947,8 +1947,8 @@ In addition to that, Resolution & Margin-Parameters are tested & adjusted.
             case MAP_GRAY:     ip[0] = 1; break;
             case MAP_RGBW:     ip[0] = 3; break;
             case MAP_RGB:      ip[0] = 3; break;
-            case MAP_CMYK:     ip[0] = 4; break;
-            case MAP_CMYKGEN:  ip[0] = 4; break;
+            case MAP_CMYK:
+            case MAP_CMYKGEN:  ip[0] = 4; color_info.polarity = GX_CINFO_POLARITY_SUBTRACTIVE; break;
             case MAP_RGBOV:    ip[0] = 3; break;
             case MAP_RGBNOV:   ip[0] = 3; break;
             default:           ip[0] = color_info.num_components; break;
@@ -1959,14 +1959,15 @@ In addition to that, Resolution & Margin-Parameters are tested & adjusted.
          case MAP_GRAY:     ncomp = 1; break;
          case MAP_RGBW:     ncomp = 4; break;
          case MAP_RGB:      ncomp = 3; break;
-         case MAP_CMYK:     ncomp = 4; break;
-         case MAP_CMYKGEN:  ncomp = 4; break;
+         case MAP_CMYK:
+         case MAP_CMYKGEN:  ncomp = 4; color_info.polarity = GX_CINFO_POLARITY_SUBTRACTIVE; break;
          case MAP_RGBOV:    ncomp = 4; break;
          case MAP_RGBNOV:   ncomp = 4; break;
          default:           ncomp = ip[0]; break;
       }
       if(UPD_CMAP_MAX < ncomp) ncomp = UPD_CMAP_MAX;
 
+      color_info.max_components = ncomp;
       if(ncomp > int_a[IA_COMPBITS].size) { /* Default ComponentBits */
          UPD_MM_GET_ARRAY(udev->memory, ip2,ncomp);
          nbits = 32 / ncomp;
@@ -2042,6 +2043,8 @@ In addition to that, Resolution & Margin-Parameters are tested & adjusted.
       }                             /* Color-Ramp */
 
       udev->color_info.num_components = ip[0];
+      udev->color_info.max_components = ip[0];
+      udev->color_info.polarity       = color_info.polarity;
       udev->color_info.depth          = ip[1];
       udev->color_info.max_gray       = (gx_color_value) ip[2];
       udev->color_info.max_color      = (gx_color_value) ip[3];
-- 
2.9.1

